File: 1472-design-browser-history/1472-design-browser-history.java
import java.util.Stack;

class BrowserHistory {

  private String current = null;
  private int sizeStackBackward = 0;
  private Stack<String> stackBackward = new Stack<String>();
  private int sizeStackForward = 0;
  private Stack<String> stackForward = new Stack<String>();

  public BrowserHistory(String homepage) {
    this.current = homepage;
    stackBackward.add(homepage);
    sizeStackBackward++;
  }

  public void visit(String url) {
    // empty stack forward
    stackForward = new Stack<String>();
    sizeStackForward = 0;

    stackBackward.add(url);
    sizeStackBackward++;
    this.current = url;

  }

  public String back(int steps) {
    if (steps == 0)
      return this.current;
    if (sizeStackBackward == 1) {
      return this.current;
    }

    String retrievedValue = stackBackward.pop();
    sizeStackBackward--;

    this.current = stackBackward.peek();
    stackForward.add(retrievedValue);
    sizeStackForward++;
    return back(steps - 1);
  }

  public String forward(int steps) {
    if (steps == 0)
      return this.current;
    if (sizeStackForward == 0) {
      return this.current;
    }

    String retrievedValue = stackForward.pop();
    sizeStackForward--;

    stackBackward.add(retrievedValue);
    sizeStackBackward++;
    this.current = retrievedValue;
    return forward(steps - 1);
  }

}

File: 1472-design-browser-history/1472-design-browser-history.ts
// 11 min

class NodeBrowser {
  value: string;
  next: NodeBrowser | null;
  prev: NodeBrowser | null;

  constructor(value: string, next: NodeBrowser | null, prev: NodeBrowser | null) {
    this.value = value;
    this.next = next;
    this.prev = prev;
  }
}

class BrowserHistory {
  // origin: NodeBrowser;
  head: NodeBrowser;

  constructor(homepage: string) {
    // this.origin = new NodeBrowser(homepage, null, null);
    this.head = new NodeBrowser(homepage, null, null);
  }

  visit(url: string): void {
    this.head.next = new NodeBrowser(url, null, this.head);
    this.head = this.head.next;
  }

  back(steps: number): string {
    let counter: number = 0;
    while (counter < steps) {
      if (this.head.prev !== null) {
        this.head = this.head.prev;
        counter++;
        continue;
      }
      break;
    }
    return this.head.value;
  }

  forward(steps: number): string {
    let i = 0;
    while (i < steps) {
      if (this.head.next !== null) {
        this.head = this.head.next;
        i++;
        continue;
      }
      break;
    }
    return this.head.value;
  }
}
